Guide for "Setup Push Notifications" (#48)
notificationapi-com/docs: docs/guides/push.md

---
sidebar_position: 4
---

# Push Notifications

NotificationAPI supports Apple Push Notifications Service (APNS) and Firebase Cloud Messaging (FCM) Push Providers. Each provider has different requirements that need to be met to deliver a message to a recipient.

> ## Apple Push Notifications Service (APNS)

Reviewer: "APNs" looks like the correct writing:

https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/sending_notification_requests_to_apns

Reviewer 2: Speaking of which, do we want to have any code references to APNs styled as `apns`? I think I've been doing `apn`...

Author: I used `apn` in the serverless: 
https://github.com/notificationapi-com/serverless/blob/f8f89d12adba70820d897cb539955de75d156cb8/models/SenderPOSTRequest.json#L101

> ## Apple Push Notifications Service (APNS)

Reviewer: "Working backwards"

Also, please add FCM docs to the best of your knowledge. For code that you are not 100% sure about, write "coming soon" until Devin can fill it out.

> ### Setup

Reviewer 2: Did we want to do a general blurb talking about whats needed and why before the setup? Or maybe at each step explain it then?

In my opinion I find integrating with services better if they explain to me early what they need and why: from there it becomes clear to the user _why_ we are doing the steps outlined.

Thoughts?

1. Go to your [Apple Developer Account](https://developer.apple.com/account)
2. Click "Certificates"
3. Click "Keys"
4. Click the "+" button
5. Name the Key
6. Click "Enable" on "Apple Push Notifications Service (APNs)"
7. Click "Continue"
8. Click "Register"
9. Click "Download"
<!-- ToDo: update URL to the push setting tab -->
10. Go to the [APNS Provider Configuration](https://app.notificationapi.com/settings/push)
11. Enter the required information
12. Click "Save"

### Getting APNS Tokens

#### With a NotificationAPI Mobile SDK

> Using a NotificationAPI Mobile SDK is the best way to set this up. All NotificationAPI Mobile SDKs sync APNS tokens to NotificationAPI and will be automatically managed. This allows you to send pushes directly to a `userId` rather than APNS tokens.

Reviewer: :::info
Sending push notifications to an Apple device requires generating a special token on the device that is referred to as APNs token or device token.
:::

NotificationAPI Mobile SDKs automatically generate and sync APNs tokens with NotificationAPI servers, so you don't have to. This allows you to send push notifications directly to a `userId` from your back-end, rather than worrying about generating, storing and retrieving the APNs token.

Reviewer 2: To add, I would remove "generate" tokens: we only intercept the generated tokens and sync them. The device itself is what generates it, not us

| Mobile SDK                                                                   | APNS Token Management | Tracking Analytics |
| :--------------------------------------------------------------------------- | :-------------------: | :----------------: |

> | [iOS](https://github.com/notificationapi-com/notificationapi-ios-sdk#readme) |       Automatic       |     Automatic      |

Reviewer: Like our other SDKs, the documentation should be written here, not in the README of the repo.

---
sidebar_position: 4
---

# Push Notifications

NotificationAPI supports Apple Push Notifications Service (APNs) and Firebase Cloud Messaging (FCM) Push Providers. Each provider has different requirements that need to be met to deliver a message to a recipient.

## Apple Push Notifications Service (APNs)

### Setup

> To send a push notification to your mobile app on an Apple device, we need some credentials from your Apple Developer account. Bellow are the required steps to generate and save APNs credentials.

Reviewer 2: ```suggestion
To send a push notification to your mobile app on an Apple device, we need some credentials from your Apple Developer account. Below are the required steps to generate and save APNs credentials.
```

Typo!

1. Go to your [Apple Developer Account](https://developer.apple.com/account)
2. Click "Certificates"
3. Click "Keys"
4. Click the "+" button
5. Name the Key
6. Click "Enable" on "Apple Push Notifications Service (APNs)"
7. Click "Continue"
8. Click "Register"
9. Click "Download"

> <!-- ToDo: update URL to the push setting tab -->

Reviewer: Remove todo

10. Go to the [APNs Provider Configuration](https://app.notificationapi.com/settings/push)
11. Enter the required information
12. Click "Save"

### Managing APNs Tokens

> :::info Sending push notifications to an Apple device requires generating a special token on the device that is referred to as APNs token or device token.

Reviewer 2: ```suggestion
:::info Sending push notifications to an Apple device requires generating a special token on the device that is referred to as an APNs token or device token.
```

Flows better

:::

> NotificationAPI Mobile SDKs automatically managing and sync APNs tokens with NotificationAPI servers, so you don't have to. This allows you to send push notifications directly to a userId from your back-end, rather than worrying about generating, storing and retrieving the APNs token.

Reviewer 2: ```suggestion
NotificationAPI Mobile SDKs automatically manage and sync APNs tokens with NotificationAPI servers, so you don't have to. This allows you to send push notifications directly to a user from your back-end, rather than worrying about generating, storing and retrieving the APNs token.
```

1. Typo
2. Im wondering for "userId" if we should change it to just "user" or if we are directly talking about code then we should use it like `userId`? Im good with whatever solution here, just offering alternatives

| Mobile SDK                                                | APNs Token Management | Tracking Analytics |
| :-------------------------------------------------------- | :-------------------: | :----------------: |

> | [iOS](https://docs.notificationapi.com/reference/ios_sdk) |       Automatic       |     Automatic      |

Reviewer: We don't do "tracking analytics" yet

## Firebase Cloud Messaging (FCM)

### Setup

> To send a push notification to your mobile app on an Android device, we need some credentials from your Firebase Cloud Messaging account. Bellow are the required steps to generate and save FCM credentials.

Reviewer 2: ```suggestion
To send a push notification to your mobile app on an Android device, we need some credentials from your Firebase Cloud Messaging account. Below are the required steps to generate and save FCM credentials.
```

Typo!

1. Create a Firebase project [here](https://firebase.google.com/)
2. Go to Project Settings
3. Click "Service Accounts"
4. Click "Generate new private key"
5. Copy the private key JSON
6. Paste the private key JSON into the [FCM Provider Configuration](https://app.notificationapi.com/settings/push)

7. Click "Save"

> :::info Sending push notifications to an Android device requires generating a special token on the device that is referred to as FCM token or device token.

Reviewer 2: ```suggestion
:::info Sending push notifications to an Android device requires generating a special token on the device that is referred to as a FCM token or device token.
```

Flows better

:::

> NotificationAPI Mobile SDKs automatically managing and sync FCM tokens with NotificationAPI servers, so you don't have to. This allows you to send push notifications directly to a userId from your back-end, rather than worrying about generating, storing and retrieving the FCM token.

Reviewer 2: ```suggestion
NotificationAPI Mobile SDKs automatically manage and sync FCM tokens with NotificationAPI servers, so you don't have to. This allows you to send push notifications directly to a userId from your back-end, rather than worrying about generating, storing and retrieving the FCM token.
```

> | Mobile SDK                                                        | APNs Token Management | Tracking Analytics |

Reviewer: Remove "tracking analytics"

Reviewer: Probably remove these tables.

Author: There is a link in it I want to keep.

| :---------------------------------------------------------------- | :-------------------: | :----------------: |
| [Android](https://docs.notificationapi.com/reference/android_sdk) |       Automatic       |     Automatic      |

> ## Sending Messages with more customizations

Reviewer 2: ```suggestion
## Sending Customized Messages
```
I think this reads better

> To add more customizations to your push notification please check out the `options.apn` attribute in the [`send`](../reference/server.md#options-additional-customization) request.

Reviewer 2: ```suggestion
To further customize your push notification please check out the `options.apn` attribute in the [`send`](../reference/server.md#options-additional-customization) request.
```

Again I think this reads better
